queue/group_remote.go

```diff
@@ -0,0 +1,387 @@
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	colls, err := getExistingCollections(ctx, g.client, g.mongooptions.DB, g.prefix)
+	if err != nil {
+		return errors.Wrap(err, "problem getting collections")
+	}
+	for i, coll := range colls {
+		// This is an optimization. If we've added to the queue recently enough, there's no
+		// need to query its contents, since it cannot be old enough to prune.
+		if t, ok := g.ttlMap[coll]; ok && time.Since(t) < g.ttl {
+			remove(colls, i)
+		}
+	}
+	catcher := grip.NewBasicCatcher()
+	wg := &sync.WaitGroup{}
+	collsDeleteChan := make(chan string, len(colls))
+	collsDropChan := make(chan string)
+
+	go func() {
+		for _, coll := range colls {
+			collsDropChan <- coll
+		}
+		close(collsDropChan)
+	}()
+	wg = &sync.WaitGroup{}
+	for i := 0; i < runtime.NumCPU(); i++ {
+		wg.Add(1)
+		go func(ch chan string) {
+			defer recovery.LogStackTraceAndContinue("panic in pruning collections")
+			defer wg.Done()
+			for nextColl := range collsDropChan {
+				c := g.client.Database(g.mongooptions.DB).Collection(nextColl)
+				one := c.FindOne(ctx, bson.M{"status.mod_ts": bson.M{"$gte": time.Now().Add(-g.ttl)}})
+				if err := one.Err(); err != nil {
+					catcher.Add(err)
+					return
+				}
+				if err := one.Decode(struct{}{}); err != nil {
+					if err != mongo.ErrNoDocuments {
+						catcher.Add(err)
+						return
+					}
+				} else {
+					return
+				}
+				one = c.FindOne(ctx, bson.M{"status.completed": false})
+				if err := one.Err(); err != nil {
+					catcher.Add(err)
+					return
+				}
+				if err := one.Decode(struct{}{}); err != nil {
+					if err != mongo.ErrNoDocuments {
+						catcher.Add(err)
+						return
+					}
+				} else {
+					return
+				}
+				if queue, ok := g.queues[idFromCollection(g.prefix, nextColl)]; ok {
+					queue.Runner().Close()
+					ch <- idFromCollection(g.prefix, nextColl)
+				}
+				if err := c.Drop(ctx); err != nil {
+					catcher.Add(err)
+				}
+			}
+		}(collsDeleteChan)
+	}
+	wg.Wait()
+	close(collsDeleteChan)
+	for id := range collsDeleteChan {
+		delete(g.queues, id)
+		delete(g.ttlMap, id)
+	}
+
+	// Another prune may have gotten to the collection first, so we should close the queue.
+	queuesDeleteChan := make(chan string, len(g.queues))
+	wg = &sync.WaitGroup{}
+outer:
+	for id, q := range g.queues {
+		for _, coll := range colls {
+			if id == idFromCollection(g.prefix, coll) {
+				continue outer
+			}
+		}
+		wg.Add(1)
+		go func(queueID string, ch chan string, qu amboy.Queue) {
+			defer recovery.LogStackTraceAndContinue("panic in pruning queues")
+			defer wg.Done()
+			qu.Runner().Close()
+			ch <- queueID
+		}(id, queuesDeleteChan, q)
+	}
+	wg.Wait()
+	close(queuesDeleteChan)
+	for id := range queuesDeleteChan {
+		delete(g.queues, id)
+		delete(g.ttlMap, id)
+	}
+	return catcher.Resolve()
+}
+
+// Close the queues.
+func (g *remoteQueueGroup) Close(ctx context.Context) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	g.canceler()
```

Comment:
is it worth nil checking this?

Reply:
you can't get a nil `remoteQueueGroup` (unless you don't error check the constructor), so I don't think so